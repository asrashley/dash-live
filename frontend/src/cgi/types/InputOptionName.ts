export interface InputOptionName {
    shortName: string;
    fullName: string;
    cgiName: string;
}