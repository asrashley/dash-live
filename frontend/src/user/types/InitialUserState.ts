export type InitialUserState = {
    pk?: number;
    username?: string;
    email?: string;
    lastLogin: string | null;
    mustChange: boolean;
    groups: string[];
};
