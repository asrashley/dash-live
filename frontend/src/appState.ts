import { createContext } from 'preact';
import { signal, computed, type Signal, type ReadonlySignal } from "@preact/signals";
import { DialogState } from './types/DialogState';
import { InitialUserState, UserState } from './types/UserState';

export interface AppStateType {
  backdrop: ReadonlySignal<boolean>;
  dialog: Signal<DialogState>;
  user: ReadonlySignal<UserState>;
  setUser: (ius: InitialUserState) => void;
}

export const AppStateContext = createContext<AppStateType>(null);

export interface CreateAppStateProps {
  userInfo: InitialUserState
}

export function createAppState(userInfo: InitialUserState = {isAuthenticated: false, groups:[]}): AppStateType {
  const dialog = signal(null);
  const userData = signal<InitialUserState>(userInfo);
  const user = computed<UserState>(() => ({
    ...userData.value,
    permissions: {
      admin: userData.value.groups.includes('ADMIN'),
      media: userData.value.groups.includes('MEDIA'),
      user: userData.value.groups.includes('USER'),
    },
  }));

  const backdrop = computed(() => {
    return dialog.value?.backdrop === true;
  });

  const setUser = (ius: InitialUserState) => {
    userData.value = {...ius};
  };

  return { dialog, backdrop, user, setUser };
}
