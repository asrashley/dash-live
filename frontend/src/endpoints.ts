import { createContext } from 'preact';
import log from 'loglevel';

import { routeMap, uiRouteMap} from '@dashlive/routemap';
import { CsrfTokenStore } from './CsrfTokenStore';
import { CsrfTokenCollection } from './types/CsrfTokenCollection';
import { AllManifests } from './types/AllManifests';
import { AllStreamsJson, AllStreamsResponse } from './types/AllStreams';
import { JWToken } from './types/JWToken';
import { MultiPeriodStream, MultiPeriodStreamJson } from './types/MultiPeriodStream';
import { DecoratedMultiPeriodStream } from "./types/DecoratedMultiPeriodStream";
import { MultiPeriodStreamSummary } from './types/MultiPeriodStreamSummary';
import { ContentRolesMap } from './types/ContentRolesMap';
import { ModifyMultiPeriodStreamJson, ModifyMultiPeriodStreamResponse } from './types/ModifyMultiPeriodStreamResponse';
import { InitialApiTokens } from './types/InitialApiTokens';
import { LoginRequest } from './types/LoginRequest';
import { LoginResponse } from './types/LoginResponse';
import { MultiPeriodStreamValidationRequest, MultiPeriodStreamValidationResponse } from './types/MpsValidation';
import { CgiOptionDescription } from './types/CgiOptionDescription';

type TokenStoreCollection = {
  files: CsrfTokenStore;
  kids: CsrfTokenStore;
  login: CsrfTokenStore;
  streams: CsrfTokenStore;
  upload: CsrfTokenStore;
};

export type ApiRequestOptions = {
  authorization?: string;
  body: RequestInit['body'];
  service: keyof TokenStoreCollection;
  signal?: AbortSignal;
  method: RequestInit['method'];
  rejectOnError?: boolean;
  query?: URLSearchParams;
};

export type GetAllStreamsProps = ApiRequestOptions & {
  withDetails: boolean
};

type RefreshAccessTokenResponse = {
  accessToken?: JWToken;
  csrfTokens: CsrfTokenCollection;
  ok: boolean;
  status: number;
};

export interface ApiRequestsProps extends Readonly<InitialApiTokens> {
  navigate: (url: string) => void;
}

export class ApiRequests {
  private csrfTokens: TokenStoreCollection;
  private accessToken: JWToken | null;
  private refreshToken: JWToken | null;
  private navigate: ApiRequestsProps['navigate'];

  constructor({csrfTokens, accessToken, refreshToken, navigate}: ApiRequestsProps) {
    this.csrfTokens = {
      files: new CsrfTokenStore(csrfTokens?.files),
      kids: new CsrfTokenStore(csrfTokens?.kids),
      login: new CsrfTokenStore(csrfTokens?.login),
      streams: new CsrfTokenStore(csrfTokens?.streams),
      upload: new CsrfTokenStore(csrfTokens?.upload),
    }
    this.accessToken = accessToken;
    this.refreshToken = refreshToken;
    this.navigate = navigate;
  }

  getAllManifests(options: Partial<ApiRequestOptions> = {}): Promise<AllManifests> {
    return this.sendApiRequest<AllManifests>(routeMap.listManifests.url(), options);
  }

  getContentRoles(options: Partial<ApiRequestOptions> = {}): Promise<ContentRolesMap> {
    return this.sendApiRequest(routeMap.contentRoles.url(), options);
  }

  getCgiOptions(options: Partial<ApiRequestOptions> = {}) : Promise<CgiOptionDescription[]> {
    return this.sendApiRequest(routeMap.cgiOptions.url(), options);
  }

  async getUserInfo(signal: AbortSignal): Promise<LoginResponse | Response> {
    if (!this.refreshToken) {
      throw new Error('Cannot get user info without a refresh token');
    }
    const options: Partial<ApiRequestOptions> = {
      authorization: this.refreshToken.jwt,
      method: 'GET',
      rejectOnError: false,
      signal,
    };
    const response = await this.sendApiRequest<LoginResponse | Response>(routeMap.login.url(), options);
    if (response['success']) {
      this.accessToken = (response as LoginResponse).accessToken;
    }
    return response;
  }

  async loginUser(request: LoginRequest, options: Partial<GetAllStreamsProps> = {}): Promise<LoginResponse> {
    const response: LoginResponse = await this.sendApiRequest<LoginResponse>(routeMap.login.url(), {
      body: JSON.stringify(request),
      method: 'POST',
      service: 'login',
      ...options,
    });
    if (response.success) {
      if (response.accessToken) {
        this.accessToken = response.accessToken;
      }
      if (response.refreshToken) {
        this.refreshToken = response.refreshToken;
      }
    }
    return response;
  }

  async logoutUser(options: Partial<GetAllStreamsProps> = {}): Promise<Response> {
    const response: Response = await this.sendApiRequest(routeMap.login.url(), {
      method: 'DELETE',
      service: 'login',
      ...options,
    });
    this.accessToken = null;
    this.refreshToken = null;
    return response;
  }

  async getAllStreams(options: Partial<GetAllStreamsProps> = {}): Promise<AllStreamsResponse> {
    const { streams, keys } = await this.sendApiRequest<AllStreamsJson>(routeMap.listStreams.url(), options);
    return { streams, keys};
  }

  getAllMultiPeriodStreams(options: Partial<ApiRequestOptions> = {}): Promise<MultiPeriodStreamSummary[]> {
    return this.sendApiRequest<MultiPeriodStreamSummary[]>(routeMap.listMps.url(), options);
  }

  async getMultiPeriodStream(mps_name: string, options: Partial<ApiRequestOptions> = {}): Promise<MultiPeriodStream> {
    const { model } = await this.sendApiRequest<MultiPeriodStreamJson>(routeMap.editMps.url({mps_name}), options);
    return model;
  }

  async addMultiPeriodStream(
      data: DecoratedMultiPeriodStream,
      options: Partial<ApiRequestOptions> = {}): Promise<ModifyMultiPeriodStreamResponse> {
    const service = 'streams';
    const csrf_token = await this.csrfTokens[service].getToken(
      options?.signal, this.getCsrfTokens);
    const { errors, success, model } = await this.sendApiRequest<ModifyMultiPeriodStreamJson>(routeMap.addMps.url(), {
      ...options,
      service,
      body: JSON.stringify({
        ...data,
        csrf_token
      }),
      method: 'PUT',
    });
    return { errors, success, model };
  }

  async modifyMultiPeriodStream(mps_name: string, data: DecoratedMultiPeriodStream,
     options: Partial<ApiRequestOptions> = {}): Promise<ModifyMultiPeriodStreamJson> {
    const service = 'streams';
    const csrf_token = await this.csrfTokens[service].getToken(
      options?.signal, this.getCsrfTokens);
    return await this.sendApiRequest<ModifyMultiPeriodStreamJson>(routeMap.editMps.url({mps_name}), {
      ...options,
      service,
      body: JSON.stringify({...data, csrf_token}),
      method: 'POST',
    });
  }

  async deleteMultiPeriodStream(mps_name: string, options: Partial<ApiRequestOptions> = {}): Promise<Response> {
    const service = 'streams';
    const csrf_token = await this.csrfTokens[service].getToken(
      options?.signal, this.getCsrfTokens
    );
    const query = new URLSearchParams({
      csrf_token,
    });
    return this.sendApiRequest<Response>(routeMap.editMps.url({mps_name}), {
      ...options,
      service,
      query,
      method: 'DELETE',
    });
  }

  async validateMultiPeriodStream(data: MultiPeriodStreamValidationRequest,
        options: Partial<ApiRequestOptions> = {}): Promise<MultiPeriodStreamValidationResponse> {
    const service = 'streams';
    const csrf_token = await this.csrfTokens[service].getToken(
      options?.signal, this.getCsrfTokens);
    return this.sendApiRequest<MultiPeriodStreamValidationResponse>(routeMap.validateMps.url(), {
      ...options,
      method: 'POST',
      service: 'streams',
      body: JSON.stringify({...data, csrf_token}),
    });
  }

  private async sendApiRequest<T>(url: string, options: Partial<ApiRequestOptions>): Promise<T> {
    const { authorization, body, service, signal, method='GET',
      rejectOnError = true } = options;
    let { query } = options;
    let usedAccessToken = false;
    const headers: Headers = new Headers({
      "Content-Type": "application/json",
    });
    if (authorization !== undefined) {
      headers.set('Authorization', `Bearer ${authorization}`);
    } else {
      if (!this.accessToken && this.refreshToken) {
        await this.getAccessToken(signal);
      }
      if (this.accessToken) {
        headers.set('Authorization', `Bearer ${this.accessToken.jwt}`);
        usedAccessToken = true;
      } else if (service) {
        const token = await this.csrfTokens[service].getToken(signal);
        if (query === undefined) {
          query = new URLSearchParams({csrf_token: token});
        } else {
          query.set('csrf_token', token);
        }
      }
    }
    if (query) {
      url = `${url}?${query.toString()}`;
    }
    if (signal && signal.aborted) {
      throw signal.reason;
    }
    const cache = 'no-cache';
    const credentials = 'same-origin';
    const mode = 'same-origin';
    url = new URL(url, document.location.origin).href;
    log.trace(`sendApiRequest(${url})`);
    let fetchResult = await fetch(url, {
      cache,
      credentials,
      mode,
      body,
      headers,
      method,
      signal,
    });
    log.trace(`url=${url} status=${fetchResult.status} usedAccessToken=${usedAccessToken}`);
    if (fetchResult.status === 401 && usedAccessToken && this.refreshToken) {
      await this.getAccessToken(signal);
      headers.set('Authorization', `Bearer ${this.accessToken.jwt}`);
      log.trace(`retrying sendApiRequest(${url})`);
      fetchResult = await fetch(url, {
          cache,
          credentials,
          mode,
          body,
          headers,
          method,
          signal,
      });
    }
    if (!fetchResult.ok) {
      if (rejectOnError) {
        throw new Error(`${ url }: ${ fetchResult.status }`);
      } else {
        return fetchResult as T;
      }
    }
    if (fetchResult.status !== 200) {
      return fetchResult as T;
    }
    const data = await fetchResult.json();
    if (signal?.aborted) {
      throw signal.reason;
    }
    if (typeof(data?.csrfTokens) === "object") {
      this.updateCsrfTokens(data.csrfTokens);
    } else if(typeof(data?.csrf_tokens) === "object") {
      this.updateCsrfTokens(data.csrf_tokens);
    } else if (typeof(data?.csrf_token) === "string") {
      this.updateCsrfTokens({[service]: data.csrf_token});
    }

    return data as T;
  }

  private updateCsrfTokens(tokens: Partial<CsrfTokenCollection>) {
    for (const [key, value] of Object.entries(tokens)) {
      if (this.csrfTokens[key] === undefined) {
        this.csrfTokens[key] = new CsrfTokenStore(value);
      } else {
        this.csrfTokens[key].setToken(value);
      }
    }
  }

  private async getAccessToken(signal: AbortSignal): Promise<JWToken> {
    if (!this.refreshToken) {
      throw new Error('Cannot request an access token without a refresh token');
    }
    const options = {
      authorization: this.refreshToken.jwt,
      rejectOnError: false,
      signal,
    };
    const data = await this.sendApiRequest<RefreshAccessTokenResponse>(routeMap.refreshAccessToken.url(), options);
    const { accessToken, csrfTokens, ok, status } = data ?? {};
    if (ok === false && status === 401) {
      this.navigate(uiRouteMap.login.url());
    }
    if (!accessToken) {
      throw new Error('Failed to refresh access token');
    }
    this.accessToken = accessToken;
    if (csrfTokens) {
      this.updateCsrfTokens(csrfTokens);
    }
    return accessToken;
  }

  private getCsrfTokens = async (signal: AbortSignal): Promise<void> => {
    if (!this.accessToken && this.refreshToken) {
      await this.getAccessToken(signal);
    }
    if (!this.accessToken) {
      throw new Error('Cannot request CSRF tokens without an access token');
    }
    const options = {
      authorization: this.accessToken.jwt,
      signal,
    };
    const data = await this.sendApiRequest<RefreshAccessTokenResponse>(routeMap.refreshCsrfTokens.url(), options);
    const { csrfTokens } = data ?? {};
    if (csrfTokens) {
      this.updateCsrfTokens(csrfTokens);
    }
  };
}

export const EndpointContext = createContext<ApiRequests>(null);
