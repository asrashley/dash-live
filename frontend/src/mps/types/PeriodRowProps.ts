import { MpsPeriod } from "../../types/MpsPeriod";

export interface PeriodRowProps {
  className?: string;
  index: number;
  item: MpsPeriod;
}
