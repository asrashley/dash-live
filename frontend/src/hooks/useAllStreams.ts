import { createContext } from "preact";
import { useContext, useEffect } from "preact/hooks";
import { useSignal, useComputed, type Signal } from "@preact/signals";

import { EndpointContext } from "../endpoints";
import { Stream } from "../types/Stream";
import { StreamTrack } from "../types/StreamTrack";
import { DecoratedStream } from "../types/DecoratedStream";

function findStreamTracks(stream: Stream): StreamTrack[] {
  const tracks = new Map();
  for (const mf of stream.media_files) {
    const { track_id, content_type, codec_fourcc } = mf;
    let track: StreamTrack = tracks.get(track_id);
    if (track === undefined) {
      track = {
        clearBitrates: 0,
        encryptedBitrates: 0,
        track_id,
        content_type,
        codec_fourcc,
      };
      tracks.set(track_id, track);
    }
    if (mf.encrypted) {
      track.encryptedBitrates += 1;
    } else {
      track.clearBitrates += 1;
    }
  }
  const ids = [...tracks.keys()];
  ids.sort();
  return ids.map((tid) => tracks.get(tid));
}

export function decorateAllStreams(streams: Stream[]): DecoratedStream[] {
  return streams.map((stream) => {
    const decoratedStream: DecoratedStream = {
      ...stream,
      tracks: findStreamTracks(stream),
    };
    return decoratedStream;
  });
}

export interface UseAllStreamsHook {
  allStreams: Signal<DecoratedStream[]>,
  loaded: Signal<boolean>,
  streamsMap: Signal<Map<string, DecoratedStream>>,
  error: Signal<string | null>;
}

export const AllStreamsContext = createContext<UseAllStreamsHook>(null);

export function useAllStreams(): UseAllStreamsHook {
  const apiRequests = useContext(EndpointContext);
  const streams = useSignal<DecoratedStream[] | undefined>();
  const loaded = useSignal<boolean>(false);
  const error = useSignal<string | null>(null);
  const allStreams = useComputed<DecoratedStream[]>(() => streams.value ?? []);
  const streamsMap = useComputed<Map<string, DecoratedStream>>(() => {
    const rv = new Map();
    if (allStreams.value) {
      for (const stream of allStreams.value) {
        rv.set(stream.pk, stream);
      }
    }
    return rv;
  });

  useEffect(() => {
    const controller = new AbortController();
    const { signal } = controller;

    const fetchAllStreamsIfRequired = async () => {
      if (!streams.value) {
        try {
          const data = await apiRequests.getAllStreams({
            signal,
            withDetails: true,
          });
          if (!signal.aborted) {
            streams.value = decorateAllStreams(data.streams);
            error.value = null;
            loaded.value = true;
          }
        } catch (err) {
          if (!signal.aborted) {
            console.error(err);
            error.value = `${err}`;
          }
        }
      }
    };

    fetchAllStreamsIfRequired();

    return () => {
      controller.abort();
    };
  }, [apiRequests, error, loaded, streams]);

  return { allStreams, loaded, streamsMap, error };
}
