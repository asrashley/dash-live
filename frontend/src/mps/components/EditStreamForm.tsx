import { useComputed } from "@preact/signals";
import { useContext, useCallback } from "preact/hooks";
import { useLocation } from "wouter-preact";

import { uiRouteMap } from "@dashlive/routemap";

import { FormRow } from "../../form/components/FormRow";
import { TextInputRow } from "../../form/components/TextInputRow";
import { useMessages } from "../../hooks/useMessages";
import { MultiPeriodModelContext } from "../../hooks/useMultiPeriodStream";
import { ButtonToolbar } from "./ButtonToolbar";
import { OptionsRow } from "./OptionsRow";
import { PeriodsTable } from "./PeriodsTable";
import { WhoAmIContext } from "../../user/hooks/useWhoAmI";
import { AppStateContext } from "../../appState";

export interface EditStreamFormProps {
  name: string;
  newStream: boolean;
}

export function EditStreamForm({ name, newStream }: EditStreamFormProps) {
  const setLocation = useLocation()[1];
  const { model, modified, errors, setFields, saveChanges } =
    useContext(MultiPeriodModelContext);
  const { dialog } = useContext(AppStateContext);
  const { user } = useContext(WhoAmIContext);
  const { appendMessage } = useMessages();
  const canModify = useComputed<boolean>(() => user.value.permissions.media);
  const validationClass = useComputed<string>(() => {
    if (!modified.value || !canModify.value) {
      return "";
    }
    return Object.keys(errors.value).length === 0
      ? "was-validated"
      : "has-validation";
  });
  const nameError = useComputed<string>(() => errors.value.name);
  const titleError = useComputed<string>(() => errors.value.title);

  const setName = useCallback(
    (ev: Event) => {
      setFields({ name: (ev.target as HTMLInputElement).value });
    },
    [setFields]
  );

  const setTitle = useCallback(
    (ev: Event) => {
      setFields({
        title: (ev.target as HTMLInputElement).value,
      });
    },
    [setFields]
  );

  const onSaveChanges = useCallback(async () => {
    const abortController = new AbortController();
    const { signal } = abortController;
    try {
      const success = await saveChanges({ signal });
      if (success && newStream) {
          const href = uiRouteMap.listMps.url();
          setLocation(href);
      }
    } catch(err) {
      appendMessage("warning", `Failed to save changes - ${err}`);
    }
  }, [appendMessage, newStream, saveChanges, setLocation]);

  const onDelete = useCallback(
    (ev: Event) => {
      ev.preventDefault();
      dialog.value = {
        backdrop: true,
        confirmDelete: {
          name,
          confirmed: false,
        },
      };
    },
    [dialog, name]
  );


  if (!model.value) {
    return <h3>Fetching data for stream "{name}"...</h3>;
  }

  const formRowClass = canModify.value ? "has-validation" : "";

  return (
    <div className={validationClass.value}>
      <TextInputRow
        name="name"
        label="Name"
        value={model.value.name}
        text="Unique name for this stream"
        onInput={setName}
        error={nameError}
        disabled={!canModify.value}
      />
      <TextInputRow
        name="title"
        label="Title"
        value={model.value.title}
        text="Title for this stream"
        onInput={setTitle}
        error={titleError}
        disabled={!canModify.value}
      />
      <OptionsRow name={name} canModify={canModify} />
      <FormRow className={formRowClass} name="periods" label="Periods">
        <PeriodsTable />
      </FormRow>
      <ButtonToolbar
        errors={errors}
        model={model}
        newStream={newStream}
        onSaveChanges={onSaveChanges}
        deleteStream={onDelete}
      />
    </div>
  );
}
