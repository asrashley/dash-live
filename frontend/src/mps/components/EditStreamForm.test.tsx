import { afterEach, beforeEach, describe, expect, test, vi } from "vitest";
import fetchMock from "@fetch-mock/vitest";
import { fireEvent } from "@testing-library/preact";
import userEvent from "@testing-library/user-event";
import { useContext } from "preact/hooks";
import { useLocation } from "wouter-preact";
import log from "loglevel";
import { setImmediate } from "timers";

import { routeMap, uiRouteMap } from "@dashlive/routemap";

import { ApiRequests, EndpointContext } from "../../endpoints";
import { FakeEndpoint } from "../../test/FakeEndpoint";
import { MockDashServer, mediaUser } from "../../test/MockServer";
import { renderWithProviders } from "../../test/renderWithProviders";
import { EditStreamForm, EditStreamFormProps } from "./EditStreamForm";
import { AllStreamsContext, useAllStreams } from "../../hooks/useAllStreams";
import {
  MultiPeriodModelContext,
  useMultiPeriodStream,
} from "../../hooks/useMultiPeriodStream";
import { InitialUserState } from "../../user/types/InitialUserState";
import { AppStateContext } from "../../appState";

vi.mock("wouter-preact", async (importOriginal) => {
  return {
    ...(await importOriginal()),
    useLocation: vi.fn(),
  };
});

function ConfirmDeleteButton() {
  const { dialog } = useContext(AppStateContext);
  const onClick = () => {
    console.log('click', dialog.value.confirmDelete);
    if (!dialog.value?.confirmDelete) {
      return false;
    }
    dialog.value = {
      backdrop: true,
      confirmDelete: {
        ...dialog.value.confirmDelete,
        confirmed: true,
      },
    };
  };
  if (!dialog.value?.confirmDelete) {
    return null;
  }
  return <button onClick={onClick}>Confirm Deletion</button>;
}

function AllStreams({ name, newStream }: EditStreamFormProps) {
  const allStreams = useAllStreams();
  const modelContext = useMultiPeriodStream({ name, newStream });

  return (
    <AllStreamsContext.Provider value={allStreams}>
      <MultiPeriodModelContext.Provider value={modelContext}>
        <EditStreamForm name={name} newStream={newStream} />
        <ConfirmDeleteButton />
      </MultiPeriodModelContext.Provider>
    </AllStreamsContext.Provider>
  );
}

describe("EditStreamForm component", () => {
  const needsRefreshToken = vi.fn();
  const hasUserInfo = vi.fn();
  const setLocation = vi.fn();
  const useLocationMock = vi.mocked(useLocation);
  const mps_name = "demo";
  let endpoint: FakeEndpoint;
  let server: MockDashServer;
  let api: ApiRequests;
  let userInfo: InitialUserState;

  const Wrapper = ({ name, newStream }: EditStreamFormProps) => {
    return (
      <EndpointContext.Provider value={api}>
        <AllStreams name={name} newStream={newStream} />
      </EndpointContext.Provider>
    );
  };

  beforeEach(() => {
    log.setLevel("error");
    useLocationMock.mockReturnValue([
      routeMap.editMps.url({ mps_name }),
      setLocation,
    ]);
    endpoint = new FakeEndpoint(document.location.origin);
    server = new MockDashServer({
      endpoint,
    });
    const user = server.login(mediaUser.email, mediaUser.password);
    expect(user).not.toBeNull();
    expect(user.groups).toEqual(mediaUser.groups);
    const { pk, username, email, groups, lastLogin, mustChange } = user;
    userInfo = {
      pk,
      username,
      email,
      groups,
      lastLogin,
      mustChange,
    };
    api = new ApiRequests({ needsRefreshToken, hasUserInfo });
    api.setRefreshToken(user.refreshToken);
    api.setAccessToken(user.accessToken);
  });

  afterEach(() => {
    endpoint.shutdown();
    vi.clearAllMocks();
    fetchMock.mockReset();
  });

  test("matches snapshot for an existing stream", async () => {
    expect(userInfo).toBeDefined();
    const { asFragment, findByText, findAllByText, whoAmI } =
      renderWithProviders(<Wrapper newStream={false} name={mps_name} />, {
        userInfo,
      });
    expect(whoAmI.user.value).toEqual({
      ...userInfo,
      isAuthenticated: true,
      permissions: {
        admin: false,
        user: true,
        media: true,
      },
    });
    await findByText('"europe-ntp"');
    await findAllByText("Tears of Steel");
    expect(asFragment()).toMatchSnapshot();
  });

  test.each<[string, boolean]>([
    ["edit a stream", false],
    ["create a new stream", true],
  ])("edit %s and save changes", async (_title:string, newStream: boolean) => {
    expect(userInfo).toBeDefined();
    const user = userEvent.setup();
    const { getByText, findAllByText, findBySelector, getBySelector } = renderWithProviders(
      <Wrapper newStream={newStream} name={newStream ? ".add" : mps_name} />,
      { userInfo }
    );
    if (!newStream) {
      await findAllByText("Tears of Steel");
    }
    const editProm = endpoint.addResponsePromise(
      newStream ? "put" : "post",
      newStream ? routeMap.addMps.url() : routeMap.editMps.url({ mps_name })
    );
    const nameElt = getBySelector('input[name="name"]') as HTMLInputElement;
    const titleElt = getBySelector('input[name="title"]') as HTMLInputElement;
    await user.clear(nameElt);
    await user.type(nameElt, "newname{enter}");
    await user.clear(titleElt);
    await user.type(titleElt, "title for this stream{enter}");
    if (newStream) {
      const addBtn = getByText("Add a Period") as HTMLButtonElement;
      await user.click(addBtn);
      const streamSel = await findBySelector('.period-stream select') as HTMLSelectElement;
      await user.selectOptions(streamSel, "Tears of Steel");
    }
    const btn = getByText(newStream ? "Save new stream" : "Save Changes") as HTMLButtonElement;
    expect(btn.disabled).toEqual(false);
    await user.click(btn);
    await expect(editProm).resolves.toEqual(
      expect.objectContaining({ status: 200 })
    );
    // allow the await for saveChanges() to have been processed
    await new Promise<void>(setImmediate);
    if (newStream) {
      expect(setLocation).toHaveBeenCalledWith(uiRouteMap.listMps.url());
    } else {
      expect(setLocation).not.toHaveBeenCalled();
    }
  });

  test("can delete a stream", async () => {
    expect(userInfo).toBeDefined();
    const { findByText, appState, whoAmI } = renderWithProviders(
      <Wrapper newStream={false} name={mps_name} />,
      { userInfo }
    );
    expect(whoAmI.user.value).toEqual({
      ...userInfo,
      isAuthenticated: true,
      permissions: {
        admin: false,
        user: true,
        media: true,
      },
    });
    expect(useLocationMock).toHaveBeenCalled();
    await findByText('"europe-ntp"');
    expect(appState.dialog.value).toBeNull();
    const btn = (await findByText("Delete Stream")) as HTMLButtonElement;
    fireEvent.click(btn);
    expect(appState.dialog.value).toEqual({
      backdrop: true,
      confirmDelete: {
        confirmed: false,
        name: mps_name,
      },
    });
  });
});
