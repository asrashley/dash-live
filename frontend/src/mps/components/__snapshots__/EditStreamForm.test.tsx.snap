// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EditStreamForm component > matches snapshot for an existing stream 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <div
      class="row mb-2 form-group "
    >
      <label
        class="col-2 col-form-label"
        for="model-name"
      >
        Name:
      </label>
      <div
        class="col-7"
      >
        <input
          class="form-control is-valid"
          id="field-name"
          name="name"
          type="text"
        />
      </div>
      <div
        class="col-3 form-text"
      >
        Unique name for this stream
      </div>
    </div>
    <div
      class="row mb-2 form-group "
    >
      <label
        class="col-2 col-form-label"
        for="model-title"
      >
        Title:
      </label>
      <div
        class="col-7"
      >
        <input
          class="form-control is-valid"
          id="field-title"
          name="title"
          type="text"
        />
      </div>
      <div
        class="col-3 form-text"
      >
        Title for this stream
      </div>
    </div>
    <div
      class="row mb-2 form-group "
    >
      <label
        class="col-2 col-form-label"
        for="model-options"
      >
        Stream Options:
      </label>
      <div
        class="col-10"
      >
        <div
          class="d-flex flex-row"
        >
          <div
            class="json bg-secondary-subtle border border-dark-subtle flex-fill me-1"
          >
            <span
              class="key"
            >
              ntps
            </span>
            : 
            <span
              class="value"
            >
              "europe-ntp"
            </span>
            , 
            <span
              class="key"
            >
              st
            </span>
            : 
            <span
              class="value"
            >
              true
            </span>
          </div>
          <button
            class="btn btn-primary"
          >
            Options
          </button>
        </div>
      </div>
    </div>
    <div
      class="row mb-2 form-group has-validation"
    >
      <label
        class="col-2 col-form-label"
        for="model-periods"
      >
        Periods:
      </label>
      <div
        class="col-10"
      >
        <div>
          <div
            class="period-table border"
          >
            <div
              class="row bg-secondary text-white table-head"
            >
              <div
                class="col period-ordering"
              >
                #
              </div>
              <div
                class="col period-id"
              >
                ID
              </div>
              <div
                class="col period-stream"
              >
                Stream
              </div>
              <div
                class="col period-start"
              >
                Start Time
              </div>
              <div
                class="col period-duration"
              >
                Duration
              </div>
              <div
                class="col period-tracks"
              >
                Tracks
              </div>
            </div>
            <ul>
              <li
                class="row mt-1 p-1 bg-secondary-subtle "
              >
                <div
                  class="col period-ordering"
                >
                  <div
                    class="dropdown"
                  >
                    <a
                      aria-expanded="false"
                      class=" dropdown-toggle"
                      href="#"
                      role="button"
                    >
                      <span
                        aria-label="three-dots icon"
                        class="bi icon bi-three-dots "
                        role="img"
                      />
                    </a>
                    <ul
                      class="dropdown-menu "
                    >
                      <li
                        data-testid="ddi_0"
                      >
                        <a
                          class="dropdown-item"
                          href="#"
                        >
                          Add another Period
                        </a>
                      </li>
                      <li
                        data-testid="ddi_1"
                      >
                        <a
                          class="dropdown-item"
                          href="#"
                        >
                          Delete Period
                        </a>
                      </li>
                    </ul>
                  </div>
                </div>
                <div
                  class="col period-id"
                >
                  <input
                    class="form-control is-valid"
                    id="field-pid_2"
                    name="pid_2"
                    required=""
                    type="text"
                  />
                </div>
                <div
                  class="col period-stream"
                >
                  <select
                    class="form-select is-valid"
                    name="stream_2"
                    required=""
                  >
                    <option
                      value=""
                    >
                      --Select a stream--
                    </option>
                    <option
                      value="1"
                    >
                      Big Buck Bunny
                    </option>
                    <option
                      value="2"
                    >
                      Tears of Steel
                    </option>
                  </select>
                </div>
                <div
                  class="col period-start"
                >
                  <input
                    class="form-control is-valid"
                    min="00:00:00"
                    name="start_2"
                    step="1"
                    type="time"
                  />
                </div>
                <div
                  class="col period-duration"
                >
                  <input
                    class="form-control is-valid"
                    min="00:00:01"
                    name="duration_2"
                    step="1"
                    type="time"
                  />
                </div>
                <div
                  class="col period-tracks"
                >
                  <button
                    aria-disabled="false"
                    class="btn btn-sm m-1 btn-success"
                  >
                    2/3 tracks
                  </button>
                </div>
              </li>
              <li
                class="row mt-1 p-1 bg-primary-subtle "
              >
                <div
                  class="col period-ordering"
                >
                  <div
                    class="dropdown"
                  >
                    <a
                      aria-expanded="false"
                      class=" dropdown-toggle"
                      href="#"
                      role="button"
                    >
                      <span
                        aria-label="three-dots icon"
                        class="bi icon bi-three-dots "
                        role="img"
                      />
                    </a>
                    <ul
                      class="dropdown-menu "
                    >
                      <li
                        data-testid="ddi_0"
                      >
                        <a
                          class="dropdown-item"
                          href="#"
                        >
                          Add another Period
                        </a>
                      </li>
                      <li
                        data-testid="ddi_1"
                      >
                        <a
                          class="dropdown-item"
                          href="#"
                        >
                          Delete Period
                        </a>
                      </li>
                    </ul>
                  </div>
                </div>
                <div
                  class="col period-id"
                >
                  <input
                    class="form-control is-valid"
                    id="field-pid_1"
                    name="pid_1"
                    required=""
                    type="text"
                  />
                </div>
                <div
                  class="col period-stream"
                >
                  <select
                    class="form-select is-valid"
                    name="stream_1"
                    required=""
                  >
                    <option
                      value=""
                    >
                      --Select a stream--
                    </option>
                    <option
                      value="1"
                    >
                      Big Buck Bunny
                    </option>
                    <option
                      value="2"
                    >
                      Tears of Steel
                    </option>
                  </select>
                </div>
                <div
                  class="col period-start"
                >
                  <input
                    class="form-control is-valid"
                    min="00:00:00"
                    name="start_1"
                    step="1"
                    type="time"
                  />
                </div>
                <div
                  class="col period-duration"
                >
                  <input
                    class="form-control is-valid"
                    min="00:00:01"
                    name="duration_1"
                    step="1"
                    type="time"
                  />
                </div>
                <div
                  class="col period-tracks"
                >
                  <button
                    aria-disabled="false"
                    class="btn btn-sm m-1 btn-success"
                  >
                    2/3 tracks
                  </button>
                </div>
              </li>
            </ul>
            <div
              class="btn-toolbar"
            >
              <button
                class="btn btn-primary btn-sm m-2"
              >
                Add a Period
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="btn-toolbar"
    >
      <button
        class="btn btn-success m-2"
        disabled=""
      >
        Save Changes
      </button>
      <button
        class="btn btn-danger m-2"
      >
        Delete Stream
      </button>
      <a
        class="btn m-2 btn-primary"
        href="/multi-period-streams"
      >
        Back
      </a>
    </div>
  </div>
</DocumentFragment>
`;
